src/Language/PureScript/CoreImp/Optimizer/MagicDo.hs

```diff
@@ -73,6 +73,8 @@ magicDo = inlineST . everywhere undo . everywhereTopDown convert
   -- Remove __do function applications which remain after desugaring
   undo :: AST -> AST
   undo (Return _ (App _ (Function _ (Just ident) [] body) [])) | ident == fnName = body
+  undo (App _ (App s1 (Function s2 Nothing [] (Block ss body)) []) []) =
```

Comment:
I made a PR to this branch that rearranges the optimizations, and removes undo in favor of running convert to a fixed point. The problem was that some cases that covert inlines only arise after undo is run.